File: perf-sdk/src/rating.ts
import type { MetricName, Rating } from "./types";

// Only the metrics web-vitals does not own. LCP, INP, CLS, FCP and TTFB
// arrive already rated by the library, and keeping a second copy of those
// thresholds here would just be something to drift out of date.
//
// At or below `good` is good, above `poor` is poor, in between is
// needs-improvement — the same shape the library uses.
const THRESHOLDS: Partial<Record<MetricName, readonly [number, number]>> = {
  RESOURCE: [1000, 3000],
  LONGTASK: [100, 250],
};

export function rate(name: MetricName, value: number): Rating {
  const bounds = THRESHOLDS[name];
  // CUSTOM and ERROR have no universal scale; they are always neutral.
  if (!bounds) return "good";
  const [good, poor] = bounds;
  if (value <= good) return "good";
  if (value <= poor) return "needs-improvement";
  return "poor";
}

File: perf-cleaner/internal/clean/rating.go
package clean

// Same Web Vitals thresholds the SDK uses. They are duplicated here on
// purpose: a line whose rating is missing or unrecognised can be repaired
// instead of dropped, and the two copies are pinned together by the
// SchemaVersion contract.
var thresholds = map[string][2]float64{
	"LCP":      {2500, 4000},
	"FCP":      {1800, 3000},
	"CLS":      {0.1, 0.25},
	"INP":      {200, 500},
	"TTFB":     {800, 1800},
	"RESOURCE": {1000, 3000},
	"LONGTASK": {100, 250},
}

func rate(metric string, value float64) string {
	bounds, ok := thresholds[metric]
	// CUSTOM and ERROR have no universal scale.
	if !ok {
		return "good"
	}
	switch {
	case value <= bounds[0]:
		return "good"
	case value <= bounds[1]:
		return "needs-improvement"
	default:
		return "poor"
	}
}
